keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1094,23 +1094,93 @@ repeated evaluations if they are shared across validations.
 
 #### Secondary Authz
 
-We have general agreement to include this as a feature, but need to provide
-a concrete design.
+We will support admission control use cases requiring permission checks:
 
-kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
-as a way of doing things like:
-
-- Validate that only a user with a specific permission can set a particular
+- Validate that only a user with a specific permission can set a particular field.
+- Validate that only a controller responsible for a finalizer can remove it from the finalizers
   field.
-- Validate that only a controller responsible for a finalizer can remove it from
-  the finalizers field.
 
-This could be supported by matching criteria, or via CEL expression access, or both.
- 
-Use cases:
+To depend on an authz decision, validation expressions can use the `authorizer`
+variable, which performs authz checks for the admission request user (the same
+use as identified by `request.userInfo`), and which will be bound at evaluation
+time to an Authorizer object supporting receiver-style function overloads:
+
+| Symbol      | Type                                                                    | Description                                                                                     |
+|-------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------|
+| path        | Authorizer.(path string) -> PathCheck                                   | Defines a check for an non-resource request path (e.g. /healthz)                                |
+| check       | PathCheck.(httpRequestVerb string) -> Decision                          | Checks if the user is authorized for the HTTP request verb on the path                          |
+| resource    | Authorizer.(kind string, group string, version string) -> ResourceCheck | Defines a check for API resources                                                               |
```

Comment:
Using a wildcard for version seems like a common enough use case that it might make sense to separate it out as another function. i.e.

```
| resource    | Authorizer.(kind string, group string) -> ResourceCheck | Defines a check for API resources |
| version     | ResourceCheck.(version string) -> ResourceCheck | Scopes a ResourceCheck to a specific version |
```

Reply:
Good catch. When is version needed? I briefly looked at the authz modes and failed to find anything where version is used..

cc @benluddy

Reply:
Yeah, it's not even exposed by RBAC. Technically it is a part of the [authorization attributes interface](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go#L57) and [SubjectAccessReview API](https://github.com/kubernetes/api/blob/master/authorization/v1/types.go#L108), but I think it would require a Webhook authorizer to actually make use of it.

I'd be OK with omitting it entirely, but given that it's part of the SubjectAccessReview, maybe someone is using it.

Reply:
Thanks for clarifying!

The semantics for SubjectAccessReview are interesting. It would be possible, for example, to restrict access to the beta version of an API.  This could be useful in a controlled migration. A cluster administrator might want to identify the remaining clients of a beta API, continue to allow them to use the beta API, but ban any other clients from using it, and then as they transition the stragglers off the beta API, remove their privileges to use that API.   BTH, I wish RBAC supported version for this exact workflow.

But for such a migration to work, all authz checks would need to check for the exact version.  Do most authz checks that exist today send version or has that ship already sailed?

Reply:
Discussed this with @liggitt  (https://kubernetes.slack.com/archives/C0EN96KUY/p1675964923289599). Version is not typically provided in authz checks and per-version authz checks are very unusual, so making version optional seems like the way to go here.

Reply:
> Version is not typically provided in authz checks

To be clear, admission-driven authz checks usually specify a version of "*", meaning "is authorized against all versions". No authorizers I'm aware of differentiate authorization by version.